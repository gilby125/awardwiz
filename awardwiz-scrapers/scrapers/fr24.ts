import { ScraperMetadata } from "../../arkalis/arkalis.js"
import { AwardWizScraper } from "../awardwiz-types.js"
import { FlightRadar24Response } from "../scraper-types/fr24.js"

export const meta: ScraperMetadata = {
  name: "fr24",
}

export const runScraper: AwardWizScraper<FlightRadar24Response> = async (arkalis, query) => {
  const fr24Url = `https://api.flightradar24.com/common/v1/search.json?query=default&origin=${query.origin}&destination=${query.destination}`
  arkalis.log("Querying FlightRader24 for carriers between:", query)
  arkalis.log(`Going to ${fr24Url}`)

  arkalis.goto(fr24Url)
  const response = await arkalis.waitFor({
    "success": { type: "url", url: fr24Url }
  })
  return JSON.parse(response.response?.body)
}
