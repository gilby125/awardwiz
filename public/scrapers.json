[
  {
    "scraper": "united",
    "supportedAirlines": [
      "star alliance",
      "AC", "JP", "A3", "CA", "AI", "NZ", "NH", "OZ", "OS", "AV", "SN", "CM", "OU", "MS", "ET", "BR", "HO", "LO", "LH", "SK", "ZH", "SQ", "SA", "LX", "TP", "TG", "TK", "UA",

      "partners",
      "EI", "VW", "EN", "AD", "4B", "9K", "WK", "EW", "HA", "HO", "OA", "3M"
    ]
  }
]