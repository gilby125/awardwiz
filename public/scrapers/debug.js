/* eslint-disable global-require */
/* eslint-disable @typescript-eslint/no-var-requires */

const puppeteer = require("puppeteer");

(async () => {
  const browser = await puppeteer.launch({ headless: false, devtools: true, defaultViewport: { width: 1920, height: 1080 } })
  const page = await browser.newPage()

  const scraper = require("./united")
  const results = await scraper({ page, context: { origin: "HNL", destination: "SFO", departureDate: "2022-04-21" } })

  debugger
})()
