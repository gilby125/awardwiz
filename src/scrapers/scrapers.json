{
  "airlineAmenities": [
    {
      "airlineCode": "AC",
      "podsAircraft": ["A330", "787", "777"]
    },
    {
      "airlineCode": "UA",
      "podsAircraft": ["787", "777", "757-200", "767"]
    },
    {
      "airlineCode": "B6",
      "podsAircraft": ["Mint", "A321"]
    },
    {
      "airlineCode": "WN",
      "podsAircraft": []
    },
    {
      "airlineCode": "AS",
      "podsAircraft": []
    }
  ],
  "airlineGroups": {
    "staralliance": ["AC", "JP", "A3", "CA", "AI", "NZ", "NH", "OZ", "OS", "AV", "SN", "CM", "OU", "MS", "ET", "BR", "HO", "LO", "LH", "SK", "ZH", "SQ", "SA", "LX", "TP", "TG", "TK", "UA", "QK"],
    "skyteam": ["AR", "AM", "UX", "AF", "CI", "MU", "OK", "DL", "GA", "AZ", "KQ", "KL", "KE", "ME", "SV", "RO", "VN", "MF"]
  },
  "saverBookingClasses": {
    "staralliance": ["X", "I", "O"],
    "AA": ["T", "U", "X", "Z"],
    "HA": ["T", "D"],
    "B6": ["I", "P"]
  },
  "excludeAirlines": ["1I", "FX", "KH", "5X", "8C", "KZ", "5Y", "OO", "K4", "NC"],
  "scrapers": [
    {
      "name": "aa",
      "supportedAirlines": ["AA", "UL", "S7", "RJ", "QR", "QF", "MH", "IB", "HA", "AY", "FJ", "9K", "BA", "AS", "TN", "CX"],
      "nativeAirline": "AA"
    },
    {
      "name": "aeroplan",
      "supportedAirlines": ["staralliance"],
      "nativeAirline": "AC"
    },
    {
      "name": "alaska",
      "supportedAirlines": ["AS", "EI", "AA", "BA", "CX", "DE", "LY", "EK", "FJ", "AY", "AU", "FI", "JL", "KE", "LA", "KS", "QF", "7H", "SQ"],
      "nativeAirline": "AS"
    },
    {
      "name": "delta",
      "supportedAirlines": ["skyteam"],
      "nativeAirline": "DL"
    },
    {
      "name": "jetblue",
      "supportedAirlines": ["B6"],
      "nativeAirline": "B6"
    },
    {
      "name": "skiplagged",
      "supportedAirlines": [],
      "excludeAirlines": ["WN"],
      "onlyCashFares": true,
      "nativeAirline": null
    },
    {
      "name": "southwest",
      "supportedAirlines": ["WN"],
      "nativeAirline": "WN"
    },
    {
      "name": "united",
      "supportedAirlines": ["staralliance", "EI", "VW", "EN", "AD", "4B", "9K", "WK", "EW", "HA", "HO", "OA", "3M"],
      "nativeAirline": "UA"
    }
  ]
}
