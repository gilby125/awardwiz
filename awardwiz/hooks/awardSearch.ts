import scrapersRaw from "../../config.json"
import { runScraper } from "../helpers/runScraper.js"
import { AWFR24Response, FlightAmenities, FlightFare, FlightWithFares, ScraperResponse, SearchQuery } from "../types/scrapers.js"
export const scraperConfig: ScrapersConfig = scrapersRaw

/** NOTE: Ensure that you run `just gen-json-schemas` to generate a new schema if you edit the below. */
/** Config file that describes which scrapers are available, which airlines they support, and other airline rules */
export type ScrapersConfig = {
  /** Airline codes for airlines which should not be processed (ex. cargo airlines: `["FX"]`) */
  excludeAirlines?: string[]

  /** Chase Ultimate Rewards only supports some airlines for booking cash fares, list the ones it doesn't here (ex. WN) */
  chaseUnsupportedAirlines?: string[]

  /** Group name to array of airlines mapping (ex. `{"staralliance": ["AC", "UA"]}`) */
  airlineGroups?: { [groupName: string]: string[] }             // eslint-disable-line -- because typescript-json-schema doesnt support Record

  /** Booking codes (per airline or group) to identify as saver results (ex. `{"UA": ["I", "O"]}`) used in addition to the scraper's `isSaver` */
  saverBookingClasses?: { [airlineOrGroup: string]: string[] }  // eslint-disable-line -- because typescript-json-schema doesnt support Record

  /** Defines amenities available depending on aircraft type for an airline */
  airlineAmenities?: AirlineAmenity[]

  /** Defines scraper metadata used to select which scrapers to use for a query */
  scrapers: Scraper[]
}

/** Defines scraper metadata used to select which scrapers to use for a query */
export type Scraper = {
  /** Name of the scraper (also the filename of the scraper) */
  name: string

  /** Set to true if this scraper should not be used */
  disabled?: boolean

  /** List of airline codes or group names which should trigger to use this scraper */
  supportedAirlines: string[]

  /** Airline code that this scraper defaults to (ex. aeroplan is "AC") */
  nativeAirline?: string | null

  /** Identifies this scraper to be used for all search queries */
  cashOnlyFares?: boolean
}

/** Defines amenities available depending on aircraft type for an airline */
export type AirlineAmenity = {
  /** Airline code (ex. UA) this amenity set is for */
  airlineCode?: string

  /** Aircraft types which have pods as the first class or business seats (ex. ["777", "Mint"]). Use "*" to identify all. */
  podsAircraft?: string[]

  /** Aircraft types which have WiFi available (ex. ["777"]). Use "*" to identify all. */
  wifiAircraft?: string[]
}

///////////////////////

export type DatedRoute = { origin: string, destination: string, departureDate: string }
export type ScraperToRun = { scraperName: string, forAirlines: string[], forDatedRoute: DatedRoute }
export type AirlineRoute = { origin: string, destination: string, airlineCode: string, airlineName: string } // remove airlinename

///////////////////////

export const findAwardFlights = async (searchQuery: SearchQuery): Promise<FlightWithFares[]> => {
  const allRoutes = expandOriginsDestinations(searchQuery)
  const fr24Responses = await Promise.all(allRoutes.map(async (route) => {
    const datedRoute: DatedRoute = { origin: route.origin, destination: route.destination, departureDate: route.departureDate }
    return runScraper<AWFR24Response>("fr24", datedRoute, undefined)
  }))
  const airlineRoutes = fr24Responses.flatMap((response) => fr24ResponseToAirlineRoutes(response))
  const scrapersToRun = scrapersByAirlineRoutes(airlineRoutes, searchQuery.departureDate)
  const scraperResponses = await Promise.all(scrapersToRun.map(async (scraperToRun) => {
    const datedRoute: DatedRoute = { origin: scraperToRun.forDatedRoute.origin, destination: scraperToRun.forDatedRoute.destination, departureDate: scraperToRun.forDatedRoute.departureDate }
    return runScraper(scraperToRun.scraperName, datedRoute, undefined)
  }))

  return flightsFromScraperResponses(scraperResponses)
}

///////////////////////

export const expandOriginsDestinations = (searchQuery: SearchQuery) => {
  return searchQuery.origins.flatMap((origin) => {
    return searchQuery.destinations.map((destination) => (
      { origin, destination, departureDate: searchQuery.departureDate } as DatedRoute
    ))
  })
}

export const fr24ResponseToAirlineRoutes = (response: AWFR24Response): AirlineRoute[] => {
  const data = response.result
  if (data === undefined)
    throw new Error("Couldn't retrieve airlines serving route")
  if (data.errors)
    throw new Error(`${data.errors.message} -- ${JSON.stringify(data.errors.errors)}`)
  if (!data.result?.response.flight.data)
    return []

  return data.result.response.flight.data
    .map((item) => ({
      origin: item.airport.origin.code.iata,
      destination: item.airport.destination.code.iata,
      airlineCode: item.airline?.code.iata ?? "",
      airlineName: item.airline?.name ?? "",
    } as AirlineRoute))

    // dont take airlines who dont have a name/code
    .filter((item) => item.airlineName !== "" && item.airlineCode !== "")

    // remove duplicates
    .filter((item, index, self) =>
      self.findIndex((t) => t.origin === item.origin && t.destination === item.destination && t.airlineCode === item.airlineCode) === index)

    // remove if airline is excluded
    .filter((item) => !scraperConfig.excludeAirlines?.includes(item.airlineCode))
}

export const scrapersByAirlineRoutes = (airlineRoutes: AirlineRoute[], departureDate: string): ScraperToRun[] => {
  const results: ScraperToRun[] = []
  for (const airlineRoute of airlineRoutes) {
    const datedRoute: DatedRoute = { origin: airlineRoute.origin, destination: airlineRoute.destination, departureDate }
    const compatibleScrapers = scraperConfig.scrapers.filter((checkScraper) => doesScraperSupportAirline(checkScraper, airlineRoute.airlineCode, true))

    for (const scraper of compatibleScrapers) {
      const existingItem = results.find((item) => item.scraperName === scraper.name && item.forDatedRoute.origin === datedRoute.origin && item.forDatedRoute.destination === datedRoute.destination)
      if (existingItem) {
        if (!existingItem.forAirlines.includes(airlineRoute.airlineCode))
          existingItem.forAirlines.push(airlineRoute.airlineCode)
      } else {
        results.push({ scraperName: scraper.name, forAirlines: [airlineRoute.airlineCode], forDatedRoute: datedRoute })
      }
    }
  }
  return results
}

export const doesScraperSupportAirline = (scraper: Scraper, airlineCode: string, includeCashOnly: boolean): boolean => {
  if (scraper.disabled)
    return false
  const codes = new Set([...scraper.supportedAirlines, ...(scraper.cashOnlyFares ? [airlineCode] : [])]   // cash-only scrapers run for all airlines
    .flatMap((code) => scraperConfig.airlineGroups?.[code as keyof ScrapersConfig["airlineGroups"]] ?? code))  // expand references airline groups inplace

  return includeCashOnly ? codes.has(airlineCode) : (codes.has(airlineCode) && !scraper.cashOnlyFares)
}

////////////////////////////////

export const flightsFromScraperResponses = (scraperResponses: ScraperResponse[]): FlightWithFares[] => {
  const flightsWithFares = scraperResponses.flatMap((response) => response.result ?? [])

  const flightsIn = flightsWithFares.map((flight) => reduceFaresToBestPerCabin(flight))
  const flightsOut = mergeFlightsByFlightNo(flightsIn)
    .map((flight) => removeCashFaresFromUnsupportedAirlines(flight))
    .map((flight) => convertCashToMilesForCashOnlyScrapers(flight))
    .map((flight) => reduceFaresToBestPerCabin(flight))
    .map((flight) => calculateAmenities(flight))
    .map((flight) => calculateSaverAwards(flight))
  return flightsOut
}

const reduceFaresToBestPerCabin = (flight: FlightWithFares): FlightWithFares => {
  const scraperFares: Record<string, FlightFare | undefined> = {}
  for (const fare of flight.fares) {
    if (!scraperFares[`${fare.scraper}${fare.cabin}`] || fare.miles < scraperFares[`${fare.scraper}${fare.cabin}`]!.miles)
      scraperFares[`${fare.scraper}${fare.cabin}`] = fare
  }

  return { ...flight, fares: Object.values(scraperFares) as FlightFare[] }
}

// Merges properties of FlightWithFares into unique FlightWithFares by flightNo
const mergeFlightsByFlightNo = (flights: FlightWithFares[]) => {
  const returnFlights: FlightWithFares[] = []
  for (const flight of flights) {
    // Try to first match by flight number
    let existingFlight = returnFlights.find((returnFlight) => returnFlight.flightNo === flight.flightNo)

    // If no match, also check for codeshares
    if (!existingFlight) {
      existingFlight = returnFlights.find((returnFlight) => returnFlight.departureDateTime.slice(0, 16) === flight.departureDateTime.slice(0, 16) && returnFlight.arrivalDateTime.slice(0, 16) === flight.arrivalDateTime.slice(0, 16))

      // If the existing flight is from a cash-only scraper, take the new flight number, since other scrapers don't usually use codeshare flights
      if (existingFlight && scraperConfig.scrapers.find((scraper) => scraper.name === existingFlight?.fares[0]!.scraper)?.cashOnlyFares)
        existingFlight.flightNo = flight.flightNo
    }

    if (existingFlight) {
      for (const key in flight) {
        if (existingFlight[key as keyof FlightWithFares] === undefined) {
          // TODO: need a better way to do this
          // @ts-expect-error
          // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
          existingFlight[key] = flight[key]
        }
      }

      for (const k in flight.amenities) {
        if (existingFlight.amenities[k as keyof FlightAmenities] === undefined)
          existingFlight.amenities[k as keyof FlightAmenities] = flight.amenities[k as keyof FlightAmenities]
      }

      // Append in the fares
      existingFlight.fares.push(...flight.fares)
    } else {
      returnFlights.push(flight)
    }
  }

  return returnFlights
}

// A scraper can return cash fares for an airline (ex Southwest) which Chase Ultimate Rewards cash fares
// dont support. We keep the hinting of the flights here, but remove the fare.
const removeCashFaresFromUnsupportedAirlines = (flight: FlightWithFares): FlightWithFares => ({
  ...flight,
  fares: flight.fares.filter((fare) => {
    if (scraperConfig.scrapers.find((scraper) => scraper.name === fare.scraper)?.cashOnlyFares)
      return !scraperConfig.chaseUnsupportedAirlines?.includes(flight.flightNo.slice(0, 2))
    return true
  })
})

const convertCashToMilesForCashOnlyScrapers = (flight: FlightWithFares) => ({
  ...flight,
  fares: flight.fares.map((fare) => {
    if (!scraperConfig.scrapers.find((scraper) => scraper.name === fare.scraper)?.cashOnlyFares)
      return fare
    return { ...fare, miles: (fare.cash * 100) / 1.5, cash: 0 }
  })
})

const calculateAmenities = (flight: FlightWithFares) => {
  const amenities = scraperConfig.airlineAmenities?.find((checkAmenity) => checkAmenity.airlineCode === flight.flightNo.slice(0, 2))
  const hasPods = amenities?.podsAircraft?.some((checkText) => ((flight.aircraft ?? "").includes(checkText) || checkText === "*"))
  const hasWifi = amenities?.wifiAircraft?.some((checkText) => ((flight.aircraft ?? "").includes(checkText) || checkText === "*"))
  return { ...flight, amenities: { ...flight.amenities, hasPods: flight.amenities.hasPods ?? hasPods, hasWiFi: flight.amenities.hasWiFi ?? hasWifi } }
}

const calculateSaverAwards = (flight: FlightWithFares) => {
  const fares = flight.fares.map((fare) => {
    let isSaver = fare.isSaverFare

    // If the airline has partners that we know about in the JSON, use that as the fare's saver award true/false
    const saverBookingClasses = Object.keys(scraperConfig.saverBookingClasses ?? {}).map((key) => {
      const keys = scraperConfig.airlineGroups?.[key as keyof ScrapersConfig["airlineGroups"]] ?? [key]   // Combine in groups
      return keys.includes(flight.flightNo.slice(0, 2)) ? scraperConfig.saverBookingClasses?.[key as keyof ScrapersConfig["saverBookingClasses"]] : undefined
    }).filter((bookingClasses): bookingClasses is string[] => !!bookingClasses).flat()    // join all matches

    if (isSaver === undefined && saverBookingClasses.length > 0)
      isSaver = saverBookingClasses.includes(fare.bookingClass!)

    // If the scraper returned an airline's availability and it's not the scraper's native airline, assume it's a saver award
    const scraper = scraperConfig.scrapers.find((checkScraper) => checkScraper.name === fare.scraper)!
    if (isSaver === undefined && scraper.nativeAirline && flight.flightNo.slice(0, 2) !== scraper.nativeAirline)
      isSaver = true

    return { ...fare, isSaverFare: isSaver }
  })
  return { ...flight, fares }
}
