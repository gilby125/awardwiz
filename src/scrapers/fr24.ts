import { ScraperMetadata } from "../scraper.js"
import { AwardWizScraper } from "../types.js"
import { FlightRadar24Response } from "./samples/fr24.js"

export const meta: ScraperMetadata = {
  name: "fr24",
}

export const runScraper: AwardWizScraper<FlightRadar24Response> = async (sc, query) => {
  const fr24Url = `https://api.flightradar24.com/common/v1/search.json?query=default&origin=${query.origin}&destination=${query.destination}`
  sc.log("Querying FlightRader24 for carriers between:", query)
  sc.log(`Going to ${fr24Url}`)

  sc.browser.goto(fr24Url)
  const response = await sc.browser.waitFor({
    "success": { type: "url", url: fr24Url }
  })
  return JSON.parse(response.response?.body)
}
