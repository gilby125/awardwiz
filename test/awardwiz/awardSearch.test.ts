import { expect, describe, it, afterEach, vi } from "vitest"
import { DatedRoute, ScrapersConfig, doesScraperSupportAirline, expandOriginsDestinations, findAwardFlights, fr24ResponseToAirlineRoutes, scrapersByAirlineRoutes } from "../../awardwiz/hooks/awardSearch.js"
import scrapersConfigOrig from "../../config.json"
import { FlightRadar24Response } from "../../awardwiz/types/fr24.js"
import { existsSync, readFileSync, writeFileSync } from "node:fs"
import { FlightWithFares, ScraperResponse } from "../../awardwiz/types/scrapers.js"
const scrapersConfig = scrapersConfigOrig as ScrapersConfig
const scrapersConfigBackup = { ...scrapersConfig }

const readScraperResponseStub = <T>(path: string) => {
  if (!existsSync(`./test/awardwiz/stubs/${path}`))
    return undefined
  const data = readFileSync(`./test/awardwiz/stubs/${path}`)
  return JSON.parse(data.toString("utf-8")) as ScraperResponse<T>
}

const writeScraperResponseStub = (path: string, data: string) =>
  writeFileSync(`./test/awardwiz/stubs/_${path}`, data)

vi.mock("../../awardwiz/helpers/runScraper.js", (importOriginal) => ({
  runScraper: async (scraperName: string, datedRoute: DatedRoute) => {
    const filename = `${scraperName}-${datedRoute.origin.toLowerCase()}-${datedRoute.destination.toLowerCase()}.json`
    const scraperResponse = readScraperResponseStub<FlightWithFares[]>(filename)
    if (scraperResponse)
      return Promise.resolve(scraperResponse)

    if (import.meta.env["WRITE_MISSING_STUBS"] === "true") {
      const rs: typeof import("../../awardwiz/helpers/runScraper.js") = await importOriginal()
      const response = await rs.runScraper(scraperName, datedRoute, undefined)
      writeScraperResponseStub(filename, JSON.stringify(response))
      return Promise.resolve(response)
    }

    throw new Error(`Missing stub for ${filename}`)
  }
}))

describe("expandOriginsDestinations", () => {
  it("shouldnt expand anything if there's only one origin and destination", () => {
    const searchQuery = { origins: ["SFO"], destinations: ["LAX"], departureDate: "2023-12-25" }
    const result = expandOriginsDestinations(searchQuery)
    expect(result).toStrictEqual([
      { origin: "SFO", destination: "LAX", departureDate: "2023-12-25" }
    ])
  })

  it("should expand all permutations of origin and destination", () => {
    const searchQuery = { origins: ["SFO", "LAX"], destinations: ["JFK", "EWR"], departureDate: "2023-12-25" }
    const result = expandOriginsDestinations(searchQuery)
    expect(result).toStrictEqual([
      { origin: "SFO", destination: "JFK", departureDate: "2023-12-25" },
      { origin: "SFO", destination: "EWR", departureDate: "2023-12-25" },
      { origin: "LAX", destination: "JFK", departureDate: "2023-12-25" },
      { origin: "LAX", destination: "EWR", departureDate: "2023-12-25" }
    ])
  })
})

describe("fr24ResponseToAirlineRoutes", () => {
  afterEach(() => {
    scrapersConfig.excludeAirlines = [ ...scrapersConfigBackup.excludeAirlines ?? [] ]
  })

  it("should parse a regular response and remove dupes", () => {
    const sfoLih = readScraperResponseStub<FlightRadar24Response>("fr24-sfo-lih.json")
    expect(fr24ResponseToAirlineRoutes(sfoLih!)).toStrictEqual([{
      origin: "SFO", destination: "LIH", airlineCode: "UA", airlineName: "United Airlines"
    }])
  })

  it("should skip airlines that are excluded", () => {
    const sfoLih = readScraperResponseStub<FlightRadar24Response>("fr24-sfo-lih.json")
    scrapersConfig.excludeAirlines = ["UA"]
    expect(fr24ResponseToAirlineRoutes(sfoLih!)).toStrictEqual([])
  })

  it("should remove empty airlines names/codes", () => {
    const missingAirlineInfo = readScraperResponseStub<FlightRadar24Response>("fr24-missing-airline-info.json")
    expect(fr24ResponseToAirlineRoutes(missingAirlineInfo!)).toStrictEqual([])
  })

  it("should parse an invalid query", () => {
    const invalid = readScraperResponseStub<FlightRadar24Response>("fr24-invalid.json")
    expect(fr24ResponseToAirlineRoutes(invalid!)).toStrictEqual([])
  })
})

describe("scrapersByAirlineRoutes", () => {
  afterEach(() => {
    scrapersConfig.scrapers = { ...scrapersConfigBackup.scrapers }
  })

  it("properly selects own scraper and a partner scraper", () => {
    const airlineRoutes = [{ origin: "SFO", destination: "HNL", airlineCode: "AS", airlineName: "Alaska Airlines" }]
    scrapersConfig.scrapers = [
      { name: "alaska", supportedAirlines: ["AS"] },
      { name: "aa", supportedAirlines: ["AA", "AS"] }
    ]
    const result = scrapersByAirlineRoutes(airlineRoutes, "2023-12-25")
    expect(result.map(r => r.scraperName)).toStrictEqual(["alaska", "aa"])
  })

  it("properly selects a cash-only scraper", () => {
    const airlineRoutes = [
      { origin: "SFO", destination: "HNL", airlineCode: "AS", airlineName: "Alaska Airlines" }
    ]
    scrapersConfig.scrapers = [
      { name: "alaska", supportedAirlines: ["AS"] },
      { name: "skiplagged", supportedAirlines: [], cashOnlyFares: true },
      { name: "united", supportedAirlines: ["UA"] },
    ]
    const result = scrapersByAirlineRoutes(airlineRoutes, "2023-12-25")
    expect(result.map(r => r.scraperName)).toStrictEqual(["alaska", "skiplagged"])
    expect(result.find(r => r.scraperName === "skiplagged")?.forAirlines).toStrictEqual(["AS"])
  })

  it("uses one compatible scraper for two requests", () => {
    const airlineRoutes = [
      { origin: "SFO", destination: "HNL", airlineCode: "AS", airlineName: "Alaska Airlines" },
      { origin: "SFO", destination: "HNL", airlineCode: "AA", airlineName: "American Airlines" }
    ]
    scrapersConfig.scrapers = [
      { name: "alaska", supportedAirlines: ["AS"] },
      { name: "aa", supportedAirlines: ["AA", "AS", "BA"] },
      { name: "united", supportedAirlines: ["AC", "UA"] },
      { name: "skiplagged", supportedAirlines: [], cashOnlyFares: true },
    ]

    const result = scrapersByAirlineRoutes(airlineRoutes, "2023-12-25")
    expect(result.map(r => r.scraperName)).toStrictEqual(["alaska", "aa", "skiplagged"])
    expect(result.find(r => r.scraperName === "aa")?.forAirlines).toStrictEqual(["AS", "AA"])
  })

  it("searches for an airline in a group", () => {
    const airlineRoutes = [{ origin: "SFO", destination: "HNL", airlineCode: "UA", airlineName: "United Airlines" }]
    scrapersConfig.scrapers = [
      { name: "united", supportedAirlines: ["staralliance"] },
      { name: "delta", supportedAirlines: ["skyteam"] }
    ]
    const result = scrapersByAirlineRoutes(airlineRoutes, "2023-12-25")
    expect(result.map(r => r.scraperName)).toStrictEqual(["united"])
  })
})

describe("doesScraperSupportAirline", () => {
  afterEach(() => {
    scrapersConfig.scrapers = [ ...scrapersConfigBackup.scrapers ]
  })

  it("should support a simple scraper lookup", () => {
    scrapersConfig.scrapers = [{ name: "united", supportedAirlines: ["UA", "AC"] }]
    expect(doesScraperSupportAirline(scrapersConfig.scrapers[0]!, "UA", true)).toBeTruthy()
    expect(doesScraperSupportAirline(scrapersConfig.scrapers[0]!, "AC", true)).toBeTruthy()
    expect(doesScraperSupportAirline(scrapersConfig.scrapers[0]!, "AA", true)).toBeFalsy()
  })

  it("should support a group lookup", () => {
    scrapersConfig.scrapers = [{ name: "united", supportedAirlines: ["staralliance"] }]
    expect(doesScraperSupportAirline(scrapersConfig.scrapers[0]!, "UA", true)).toBeTruthy()
    expect(doesScraperSupportAirline(scrapersConfig.scrapers[0]!, "DL", true)).toBeFalsy()
  })

  it("should support skipping disabled scrapers", () => {
    scrapersConfig.scrapers = [
      { name: "united", supportedAirlines: ["staralliance"], disabled: true },
      { name: "delta", supportedAirlines: ["skyteam"] },
      { name: "skiplagged", supportedAirlines: [], cashOnlyFares: true, disabled: true },
    ]
    expect(doesScraperSupportAirline(scrapersConfig.scrapers[0]!, "UA", true)).toBeFalsy()
    expect(doesScraperSupportAirline(scrapersConfig.scrapers[1]!, "DL", true)).toBeTruthy()
    expect(doesScraperSupportAirline(scrapersConfig.scrapers[1]!, "UA", true)).toBeFalsy()
    expect(doesScraperSupportAirline(scrapersConfig.scrapers[2]!, "AA", true)).toBeFalsy()
  })

  it("should support cash-only fares", () => {
    scrapersConfig.scrapers = [
      { name: "skiplagged1", supportedAirlines: [], cashOnlyFares: true, disabled: true },
      { name: "skiplagged2", supportedAirlines: [], cashOnlyFares: true, disabled: false },
    ]
    expect(doesScraperSupportAirline(scrapersConfig.scrapers[0]!, "AA", true)).toBeFalsy()
    expect(doesScraperSupportAirline(scrapersConfig.scrapers[1]!, "AA", true)).toBeTruthy()
    expect(doesScraperSupportAirline(scrapersConfig.scrapers[0]!, "AA", false)).toBeFalsy()
    expect(doesScraperSupportAirline(scrapersConfig.scrapers[1]!, "AA", false)).toBeFalsy()
  })
})

describe("findAwardFlights", () => {
  it("should support a simple search", async () => {
    const flights = await findAwardFlights({origins: ["SFO", "OAK"], destinations: ["HNL"], departureDate: "2023-12-25"})
    expect(flights).toMatchSnapshot()
  })
}, { timeout: 30000 })
