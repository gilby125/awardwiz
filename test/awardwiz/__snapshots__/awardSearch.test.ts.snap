// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`findAwardFlights > should support a simple search 1`] = `
[
  {
    "amenities": {
      "hasPods": false,
      "hasWiFi": false,
    },
    "arrivalDateTime": "2023-12-25 17:23",
    "departureDateTime": "2023-12-25 13:40",
    "destination": "HNL",
    "duration": 343,
    "fares": [
      {
        "cabin": "economy",
        "cash": 0,
        "currencyOfCash": "USD",
        "isSaverFare": false,
        "miles": 45634.666666666664,
        "scraper": "skiplagged",
      },
    ],
    "flightNo": "UA 300",
    "origin": "SFO",
  },
  {
    "amenities": {
      "hasPods": false,
      "hasWiFi": false,
    },
    "arrivalDateTime": "2023-12-25 10:37",
    "departureDateTime": "2023-12-25 07:00",
    "destination": "HNL",
    "duration": 337,
    "fares": [
      {
        "cabin": "economy",
        "cash": 0,
        "currencyOfCash": "USD",
        "isSaverFare": false,
        "miles": 31968.666666666668,
        "scraper": "skiplagged",
      },
    ],
    "flightNo": "UA 1581",
    "origin": "SFO",
  },
  {
    "aircraft": "Boeing 737-900 (Winglets) Passenger",
    "amenities": {
      "hasPods": false,
      "hasWiFi": true,
    },
    "arrivalDateTime": "2023-12-25 15:40",
    "departureDateTime": "2023-12-25 11:58",
    "destination": "HNL",
    "duration": 342,
    "fares": [
      {
        "cabin": "economy",
        "cash": 0,
        "currencyOfCash": "USD",
        "isSaverFare": undefined,
        "miles": 37566.666666666664,
        "scraper": "skiplagged",
      },
      {
        "bookingClass": "C",
        "cabin": "business",
        "cash": 5.6,
        "currencyOfCash": "USD",
        "isSaverFare": false,
        "miles": 95000,
        "scraper": "alaska",
      },
      {
        "bookingClass": "H",
        "cabin": "economy",
        "cash": 5.6,
        "currencyOfCash": "USD",
        "isSaverFare": false,
        "miles": 50000,
        "scraper": "alaska",
      },
    ],
    "flightNo": "AS 877",
    "origin": "SFO",
  },
  {
    "amenities": {
      "hasPods": false,
      "hasWiFi": false,
    },
    "arrivalDateTime": "2023-12-25 12:30",
    "departureDateTime": "2023-12-25 09:00",
    "destination": "HNL",
    "duration": 330,
    "fares": [
      {
        "cabin": "economy",
        "cash": 0,
        "currencyOfCash": "USD",
        "isSaverFare": false,
        "miles": 31968.666666666668,
        "scraper": "skiplagged",
      },
    ],
    "flightNo": "UA 1175",
    "origin": "SFO",
  },
  {
    "amenities": {
      "hasPods": false,
      "hasWiFi": undefined,
    },
    "arrivalDateTime": "2023-12-25 11:40",
    "departureDateTime": "2023-12-25 08:00",
    "destination": "HNL",
    "duration": 340,
    "fares": [
      {
        "cabin": "economy",
        "cash": 0,
        "currencyOfCash": "USD",
        "isSaverFare": false,
        "miles": 22300,
        "scraper": "skiplagged",
      },
    ],
    "flightNo": "HA 11",
    "origin": "SFO",
  },
  {
    "amenities": {
      "hasPods": false,
      "hasWiFi": false,
    },
    "arrivalDateTime": "2023-12-25 22:12",
    "departureDateTime": "2023-12-25 18:40",
    "destination": "HNL",
    "duration": 332,
    "fares": [
      {
        "cabin": "economy",
        "cash": 0,
        "currencyOfCash": "USD",
        "isSaverFare": false,
        "miles": 45666.666666666664,
        "scraper": "skiplagged",
      },
    ],
    "flightNo": "UA 473",
    "origin": "SFO",
  },
  {
    "amenities": {
      "hasPods": false,
      "hasWiFi": false,
    },
    "arrivalDateTime": "2023-12-25 20:07",
    "departureDateTime": "2023-12-25 16:35",
    "destination": "HNL",
    "duration": 332,
    "fares": [
      {
        "cabin": "economy",
        "cash": 0,
        "currencyOfCash": "USD",
        "isSaverFare": false,
        "miles": 31968.666666666668,
        "scraper": "skiplagged",
      },
    ],
    "flightNo": "UA 1509",
    "origin": "SFO",
  },
  {
    "amenities": {
      "hasPods": false,
      "hasWiFi": false,
    },
    "arrivalDateTime": "2023-12-25 14:56",
    "departureDateTime": "2023-12-25 11:10",
    "destination": "HNL",
    "duration": 346,
    "fares": [
      {
        "cabin": "economy",
        "cash": 0,
        "currencyOfCash": "USD",
        "isSaverFare": false,
        "miles": 31968.666666666668,
        "scraper": "skiplagged",
      },
    ],
    "flightNo": "UA 2380",
    "origin": "SFO",
  },
  {
    "amenities": {
      "hasPods": false,
      "hasWiFi": undefined,
    },
    "arrivalDateTime": "2023-12-25 12:05",
    "departureDateTime": "2023-12-25 08:25",
    "destination": "HNL",
    "duration": 340,
    "fares": [
      {
        "cabin": "economy",
        "cash": 0,
        "currencyOfCash": "USD",
        "isSaverFare": false,
        "miles": 18966.666666666668,
        "scraper": "skiplagged",
      },
    ],
    "flightNo": "HA 47",
    "origin": "OAK",
  },
]
`;
